.npmrc

```diff
@@ -3,3 +3,4 @@ engine-strict = true
 legacy-peer-deps = true
 prefer-dedupe = true
 lockfile-version = 3
+node-options = "--max-old-space-size=8192"
```

Comment:
Adding more memory for build scripts (see https://github.com/WordPress/gutenberg/pull/67422).

Before landing this PR:
- let's discuss if there's a better way than using `.npmrc`;
- let's see if we can spot any ways to reduce the TypeScript complexity

Reply:
8 gigs is a lot 😓 we might need to spike on seeing what TS hands the most with and cleaning it up.

Reply:
At least on my machine, I tried a few lower values but the project would not reliably build. It looks like we really need to profile the TS build and understand how we can optimize it.

Are we ok to merge this change for the time being, or you do have other suggestions?

Reply:
I'm a bit concerned with merging it as-is, since it's not just CI - we might actually be making the project inaccessible for people with less capable machines. We'll need to profile what makes types so complex and memory-consuming. [This approach](https://carlrannaberg.medium.com/overcoming-javascript-heap-out-of-memory-error-during-typescript-compilation-in-a-mui5-react-21396cc8a4e1) has historically served me well.

Reply:
I did some quick profiling:

* Ran `npm run clean:package-types`
* Added profiling to the existing `tsc` command:

```diff
diff --git a/package.json b/package.json
index 07d725c4a3e..89b7b185ad5 100644
--- a/package.json
+++ b/package.json
@@ -178,7 +178,7 @@
        "scripts": {
                "build": "npm run build:packages && wp-scripts build",
                "build:analyze-bundles": "npm run build -- --webpack-bundle-analyzer",
-               "build:package-types": "node ./bin/packages/validate-typescript-version.js && ( tsc --build || ( echo 'tsc failed. Try cleaning up first: `npm run clean:package-types`'; exit 1 ) ) && node ./bin/packages/check-build-type-declaration-files.js",
+               "build:package-types": "node ./bin/packages/validate-typescript-version.js && ( tsc --build --extendedDiagnostics --generateTrace ./ts-traces || ( echo 'tsc failed. Try cleaning up first: `npm run clean:package-types`'; exit 1 ) ) && node ./bin/packages/check-build-type-declaration-files.js",
                "prebuild:packages": "npm run clean:packages && npm run --if-present --workspaces build",
                "build:packages": "npm run --silent build:package-types && node ./bin/packages/build.js",
                "postbuild:packages": " npm run --if-present --workspaces build:wp",
```
* Ran `npm run build:package-types`
* Ran ` npx @typescript/analyze-trace ts-traces > analysis.txt`

Saw this result:

```
Hot Spots
├─ Check file [35m/users/marin/projects/gutenberg/packages/components/src/menu/stories/[36mindex.story.tsx[39m[35m[39m (9432ms)
│  ├─ Check deferred node from (line 70, char 2) to (line 117, char 9) (1626ms)
│  ├─ Check deferred node from (line 305, char 10) to (line 355, char 10) (1027ms)
│  ├─ Check deferred node from (line 378, char 4) to (line 413, char 11) (1020ms)
│  ├─ Check deferred node from (line 579, char 6) to (line 613, char 13) (1019ms)
│  ├─ Check deferred node from (line 184, char 3) to (line 292, char 10) (996ms)
│  ├─ Check deferred node from (line 531, char 3) to (line 556, char 10) (988ms)
│  ├─ Check deferred node from (line 121, char 66) to (line 159, char 9) (965ms)
│  └─ Check deferred node from (line 483, char 4) to (line 497, char 11) (922ms)
```

then I imported the props types manually:

```
diff --git a/packages/components/src/menu/stories/index.story.tsx b/packages/components/src/menu/stories/index.story.tsx
index 3c66cc6bb5b..dcd890370a1 100644
--- a/packages/components/src/menu/stories/index.story.tsx
+++ b/packages/components/src/menu/stories/index.story.tsx
@@ -20,6 +20,7 @@ import Button from '../../button';
 import Modal from '../../modal';
 import { createSlotFill, Provider as SlotFillProvider } from '../../slot-fill';
 import { ContextSystemProvider } from '../../context';
+import type { MenuProps } from '../types';
 
 const meta: Meta< typeof Menu > = {
 	id: 'components-experimental-menu',
@@ -66,7 +67,7 @@ const meta: Meta< typeof Menu > = {
 };
 export default meta;
 
-export const Default: StoryFn< typeof Menu > = ( props ) => (
+export const Default: StoryFn< typeof Menu > = ( props: MenuProps ) => (
 	<Menu { ...props }>
 		<Menu.TriggerButton
 			render={ <Button __next40pxDefaultSize variant="secondary" /> }
@@ -118,7 +119,7 @@ export const Default: StoryFn< typeof Menu > = ( props ) => (
 );
 Default.args = {};
 
-export const WithSubmenu: StoryFn< typeof Menu > = ( props ) => (
+export const WithSubmenu: StoryFn< typeof Menu > = ( props: MenuProps ) => (
 	<Menu { ...props }>
 		<Menu.TriggerButton
 			render={ <Button __next40pxDefaultSize variant="secondary" /> }
@@ -162,7 +163,7 @@ WithSubmenu.args = {
 	...Default.args,
 };
 
-export const WithCheckboxes: StoryFn< typeof Menu > = ( props ) => {
+export const WithCheckboxes: StoryFn< typeof Menu > = ( props: MenuProps ) => {
 	const [ isAChecked, setAChecked ] = useState( false );
 	const [ isBChecked, setBChecked ] = useState( true );
 	const [ multipleCheckboxesValue, setMultipleCheckboxesValue ] = useState<
@@ -296,7 +297,7 @@ WithCheckboxes.args = {
 	...Default.args,
 };
 
-export const WithRadios: StoryFn< typeof Menu > = ( props ) => {
+export const WithRadios: StoryFn< typeof Menu > = ( props: MenuProps ) => {
 	const [ radioValue, setRadioValue ] = useState( 'two' );
 	const onRadioChange: React.ComponentProps<
 		typeof Menu.RadioItem
@@ -366,7 +367,7 @@ const modalOnTopOfMenuPopover = css`
 `;
 
 // For more examples with `Modal`, check https://ariakit.org/examples/menu-wordpress-modal
-export const WithModals: StoryFn< typeof Menu > = ( props ) => {
+export const WithModals: StoryFn< typeof Menu > = ( props: MenuProps ) => {
 	const [ isOuterModalOpen, setOuterModalOpen ] = useState( false );
 	const [ isInnerModalOpen, setInnerModalOpen ] = useState( false );
 
@@ -477,7 +478,7 @@ const Fill = ( { children }: { children: React.ReactNode } ) => {
 	);
 };
 
-export const WithSlotFill: StoryFn< typeof Menu > = ( props ) => {
+export const WithSlotFill: StoryFn< typeof Menu > = ( props: MenuProps ) => {
 	return (
 		<SlotFillProvider>
 			<Menu { ...props }>
@@ -525,7 +526,7 @@ const toolbarVariantContextValue = {
 		variant: 'toolbar',
 	},
 };
-export const ToolbarVariant: StoryFn< typeof Menu > = ( props ) => (
+export const ToolbarVariant: StoryFn< typeof Menu > = ( props: MenuProps ) => (
 	// TODO: add toolbar
 	<ContextSystemProvider value={ toolbarVariantContextValue }>
 		<Menu { ...props }>
@@ -560,7 +561,7 @@ ToolbarVariant.args = {
 	...Default.args,
 };
 
-export const InsideModal: StoryFn< typeof Menu > = ( props ) => {
+export const InsideModal: StoryFn< typeof Menu > = ( props: MenuProps ) => {
 	const [ isModalOpen, setModalOpen ] = useState( false );
 	return (
 		<>
```

then I ran the same commands again, and the offending file was no longer recognized as a hot spot.

Here is some comparison before and addressing the hot spot:

```
| Aggregate Types                     | 1778931  | 1758658  |
| ----------------------------------- | -------- | -------- |
| Aggregate Instantiations            | 35569209 | 35569209 |
| Aggregate Memory used               | 6360620K | 5175058K |
| Aggregate Assignability cache size  | 1639505  | 1529965  |
| Aggregate Identity cache size       | 37989    | 21925    |
| Aggregate Subtype cache size        | 157677   | 157661   |
| Aggregate Strict subtype cache size | 21925    | 21925    |
| Aggregate Tracing time              | 0.96s    | 0.61s    |
| Aggregate I/O Read time             | 0.32s    | 0.28s    |
| Aggregate Parse time                | 1.66s    | 1.45s    |
| Aggregate ResolveModule time        | 0.44s    | 0.50s    |
| Aggregate ResolveTypeReference time | 0.01s    | 0.01s    |
| Aggregate ResolveLibrary time       | 0.06s    | 0.05s    |
| Aggregate Program time              | 3.50s    | 3.22s    |
| Aggregate Bind time                 | 1.05s    | 0.88s    |
| Aggregate Check time                | 86.17s   | 70.45s   |
| Aggregate transformTime time        | 3.26s    | 3.13s    |
| Aggregate I/O Write time            | 0.39s    | 0.37s    |
| Aggregate printTime time            | 4.53s    | 4.31s    |
| Aggregate Emit time                 | 4.55s    | 4.32s    |
| Aggregate Dump types time           | 22.39s   | 19.22s   |
| Aggregate Source Map time           | 0.04s    | 0.03s    |
| Aggregate commentTime time          | 0.11s    | 0.12s    |
| Config file parsing time            | 0.08s    | 0.08s    |
| Up-to-date check time               | 0.00s    | 0.00s    |
| Build time                          | 118.87s  | 99.36s   |
```

While there are more opportunities for improving the types outside of this PR's scope, I believe this should address the primary issue and allow you to remove the increased memory flag.

Reply:
Thank you, Marin!

I wonder if rebasing to include the latest version of Storybook may also improve performance. I was also planning to migrate the Stories to CSF3 as a follow-up, but I may just do it now in a separate PR and then rebase, in case it also helps.